designs/0005-static-matrices.md

```diff
@@ -0,0 +1,252 @@
+`_varmat` to use a `varmat` type.
+
+A basic bernoulli MRP model with five population level covariates and seven group terms is available
+[here](https://github.com/bbbales2/computations_in_glms/blob/master/benchmark_model/mrp_`varmat`.stan).
+
+The initial `varmat` implementation of this model achieves a per-gradient speedup of around two.
+
+Now that `varmat` variables can be assigned, it is possible that the compiler silently replace every
+`matvar` in Stan with a `varmat`, and none of the standard Stan use cases would be affected.
+
+There are at least three major implementation strategies available at the Stanc3 level.
+
+1. Implement a new Stan type to let the user choose explicitly between `matvar` and `varmat` types
+
+2. Use `matvar` types if there is at least one function in the Stan program that does not support `varmat`
+
+3. Use `varmat` types by default, but if variable assignment is detected, or a `varmat` is used
+in an unsupported function convert to and from `matvar` automatically.
+
+This list is not exhaustive, and if there are other attractive language implementation options,
+it can be revised.
+
+The original proposal was leaning towards #1 because it most naturally corresponds to the Math
+implementation. However, this leaves the problem of explicitly declaring `varmat` to the user. This
+creates a problem two for functions defined to take matrix, vector, or row vector types. Would
+they all work with `varmat`? If not, the list of functions in Stan will get much longer to contain
+all the functions that support `varmat` and combinations of `varmat` and `matvar` arguments.
+Similarly the user's manual and function guide would get longer as well. If there is some
+under the hood conversion between `varmat` and `matvar` so that all functions do work with either
+type, then that leads to the question, why have the different types at all?
+
+The advantage of #2 is that it would allow programs that can be written using only `varmat` functions
+to be very fast (and not even worry about any conversion to `matvar` or not). The problem with this
+implementation is that, the difference between `varmat` and `matvar` would again need documented
+at the user level and some notation would need added to the user manual to indicate if a function
+is or is not available for optimization.
+
+The advantage of #3, similarly to #2, is that `varmat` and `matvar` types could be handled
+automatically by the compiler. If the user requires functions only available for `matvar`,
+there will be some slowdown as conversions between `varmat` and `matvar` are handled. However,
+if the conversions are only used in relatively cheap parts of the computation, then the expensive
+parts of the calculation can still try to take advantage of `varmat` types. The difficulty here
+is compiler implementation which would need to be more advanced to detect when it could use
+`varmat` and when it would need to convert to `matvar`.
```

Comment:
Is it possible to, as the compiler is building the program, trace through where an object is used? If so then an alt here is that when the compiler is parsing the AST, if it sees that a matrix is only going through functions supported by varmat and is not single cell accessed or read from, then the compiler can just swap out the `Eigen::Matrix<var>` with `var_value<Eigen::Matrix>`

pinging @rybern, is this a feature we have in the optimization stuff for the compiler?

Reply:
Yes, that is possible, we do that for OpenCL. This is part of transform MIR.

Reply:
But before we go that direction we need to see how much single cell access affects performance and compare that with the cost of a varmat-> matvar.

Reply:
> Yes, that is possible, we do that for OpenCL. This is part of transform MIR.

Just a sanity check, do we also check the path of the result as well? Since `var<matrix>` functions return var matrices we need to traverse the whole path for the object

> But before we go that direction we need to see how much single cell access affects performance and compare that with the cost of a varmat-> matvar.

@t4c1 would you have any time to make benchmarks for this? I might be able to later in the week. My guess is that it's a big difference

Reply:
We need to do multiple passes then if we want to traverse.

Reply:
^yeah that's what I mean, we need to have the ability to do multiple sweeps. ftr I have no idea if that's possible or reasonable in the compiler